File: go/crypto-square/crypto_square.go
package cryptosquare

import (
	"math"
	"strings"
	"unicode"
)

const testVersion = 2

func normalize(input string) (output string) {
	for _, r := range input {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			output += string(r)
		}

	}
	output = strings.ToLower(output)
	return
}

func getRectangle(length int) (row, col int) {

	// Get the approximate square side
	row = int(math.Sqrt(float64(length)))
	col = row
	for col*row < length {
		if col == row {
			col++
		} else if col > row {
			row++
		}
	}
	return
}

func Encode(input string) (output string) {
	normalized := normalize(input)
	r, c := getRectangle(len(normalized))
	for j := 0; j < c; j++ {
		for i := 0; i < r; i++ {
			if j+i*c >= len(normalized) {
				break
			}
			output += string(normalized[j+i*c])
		}
		output += " "
	}
	output = strings.Trim(output, " ")
	return
}

File: haskell/crypto-square/src/CryptoSquare.hs
module CryptoSquare
  ( encode
  ) where

import           Data.Char       (isAlphaNum, toLower)
import           Data.List       (transpose)
import           Data.List.Split

encode :: String -> String
encode xs = unwords $ transpose $ chunksOf (rect (length normalized)) normalized
  where
    normalized = normalize xs

rect :: Int -> Int
rect = ceiling . (sqrt :: Double -> Double) . fromIntegral

normalize :: String -> String
normalize = map toLower . filter isAlphaNum
